### ISSUE TITLE
Generate new history entry when taking a snapshot

### ISSUE THREAD

**Issue by elstoc** [darktable-org/darktable#14296]
I often use snapshot functionality to compare different parameter settings for a given module (for example, monochrome renderings using color calibration). However, because multiple changes to a single module do not generate a new history entry, it is not possible, after comparing snapshots, to see/select the one I've decided to use (because every snapshot is of the same numbered history item). What I'd like is to be able to compare options/settings and then, using the number assigned to the snapshot, simply click the history item I've chosen to use.

One possible solution might be that if the latest history entry has a snapshot associated with it, force a new history entry to be added when changing any module parameters.

**Comment by dterrahe** [comment]
Or double clicking the snapshot would restore the exact history used to create it.

**Comment by elstoc** [comment]
Maybe. I'm always wary about using double-click behaviour, after having seen so many people use double-click for everything even when single-click is all that's needed. I guess I still prefer the idea of being able to see the parameters for each of them -- the fact that I've created a snapshot is a good indication I want to compare it with something, and having all the information at my fingertips (via separate history items) would be nice.

**Comment by dterrahe** [comment]
These are really two separate points, but I'm responding to your wish to be able to get back to the state of the snapshot, which would _require_ an individual history item, but that would disappear if you press compress history and as far as I'm aware, there's no way to get your snapshot settings back then (apart from finding it in undo). In the old situation (snapshot = pixmap) this was understood, but in the new one, when the whole history is saved anyway, I believe it could be retrieved and put back. @turbogit?

> even when single-click is all that's needed.

Single click _shows_ the snapshot? But I agree, a separate "restore snapshot" button would be better. Although on slow machines, first having to _select_ a snapshot (which triggers the generation of the overlay image), before being able to restore it (and immediately discarding the just calculated image) can be annoying with slow modules. No easy way around that, without building in double-click delays and associated complexity.

> I still prefer the idea of being able to see the parameters for each of them

Do you mean using the history tooltip? It would be nice to have a similar snapshot tooltip that shows _all_ the differences between all the modules that are either in the current history or the snapshot history (with newly enabled/disabled only showing the non-default values). But this might be a bit of a task to implement, be slow and possibly big. It would be the means for a "total image xmp diff" that people have been requesting elsewhere (load one image or duplicate, take a snapshot, load the other one, look at snapshot diff).

**Comment by elstoc** [comment]
> Single click shows the snapshot?

I was just referring to the fact that people often don't know the difference between single-click and double-click and tend to double-click everything. I guess I'm fine with double-clicking to restore but I would maybe prefer a button so it could be done explicitly and to prevent "accidental" double-click changing actual module parameters. I would still prefer separate history entries though, because then you could just select the appropriate history entry, which keeps separation of responsibilities between the modules.

> Do you mean using the history tooltip?

No really I just meant being able to both be able to click on individual history items to see their parameters and click on the snapshot of the same number to see its impact. I feel like it would be clearer to have a 1:1 relationship between numbered snapshot and history items (obviously history compression is a complication either way).

**Comment by dterrahe** [comment]
BTW, how do we feel about long-click vs double-click? Much easier to implement (immediate action on button release, depending on how long ago clicked vs having to wait for possible next click which might not come so now we need to fire our own timeout and possibly clean it up later) and might be easier for pen and carpal as well?

**Comment by dterrahe** [comment]
> I would still prefer separate history entries though

Agreed.

> I would maybe prefer a button

Agreed too, but, as mentioned, this may introduce delay because we need "select item" + "click restore button" and the first triggers calculation. I would like _both_ the button _and_ a long click to do it _without_ having to click the separate button after waiting for a recalc. Do you have the same hesitation with long click as with double click?

> people often don't know the difference between single-click and double-click

That's why I never showed my mother dt. Well, besides all the other reasons.

> keeps separation of responsibilities between the modules.

I'm afraid I'm not getting this. Getting back to a _snapshot_ vs some random history item seems to me _exactly_ the realm of the snapshot module. The fact that it disappointingly couldn't do it in the past notwithstanding. And it may be _needed_ because the matching history item might not exist anymore if you pressed compress (or I guess undo).

**Comment by elstoc** [comment]
> BTW, how do we feel about long-click vs double-click?

I hate long-click (even more than I dislike double-click) as it's really a pain to simulate on a touchpad (I mostly use a combined touchpad-and-keyboard device) and virtually no other apps support it, so it feels like somewhat of a hidden feature.

> I'm afraid I'm not getting this. Getting back to a snapshot vs some random history item seems to me exactly the realm of the snapshot module.

I see the snapshots module as a way of _visualising_ a point in the history and comparing it with the currently-selected history item, and the history module as a way to _change the current edit and see the settings_. That's not to say the snapshot module can't be both things but I find it easier to get my head around with the responsibilities separated out like this.

**Comment by dterrahe** [comment]
> so it feels like somewhat of a hidden feature

What I meant is: you explained that the drawback of double-click is that people might use it accidentally. Do you feel there is that same risk with long click? Imho both are equally "undiscoverable" and for touchpad users etc we'd have the button. Would there be a drawback to having a "hidden" if you will feature for those who would be annoyed by the select+click button approach and who are using the hardware that allows them to be more expressive?

> That's not to say the snapshot module can't be both things

The double negatives suggest we agree that it _could_ be both things. My head wraps differently from yours sometimes ;-)

**Comment by elstoc** [comment]
> What I meant is: you explained that the drawback of double-click is that people might use it accidentally. Do you feel there is that same risk with long click?

Maybe. Given the choice I much prefer double-click to long-click.

> The double negatives suggest we agree that it could be both things

Yes. I guess I worded it that way to also include the implication that I'm not sure it _should_ be both things.

**Comment by dterrahe** [comment]
> I'm not sure it should be both things.

I got that. Which is why I remarked that our brains work differently; to me it is _obvious_ that the snapshots module, if it is used to _create_ snapshots, "should" also be able to _restore_ them (rather than requiring a manual lookup of history number). But since I'm not claiming to be any kind of authority on anything and at the moment not likely to implement this myself, my opinion is just that.
